File: DLL.java
public class DLL {

    static DLL head = null;

    int value;
    DLL next;
    DLL prev;

    public DLL(int value) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }

    public DLL(int value, DLL next, DLL prev) {
        this.value = value;
        this.next = next;
        this.prev = prev;
    }

    public static void add(int value) {
        DLL newNode = new DLL(value);
        if (head == null) {
            head = newNode;
            return;
        }

        DLL current = head;

        while (current.next != null) {
            current = current.next;
        }

        current.next = newNode;

    }

    public static void printList() {
        for (DLL current = head; current != null; current = current.next) {
            System.out.println(current.value);
        }
    }

    public static void deleteFirst(int value) {
        DLL current = head;

        while (current != null) {
            if (current.value == value) {
                current.prev.next = current.next;
            }
        }
        System.err.println("Node does not exist");
    }

    public static void insertAt(int index, int value) {
        if (head == null) {
            System.out.println("");
            return;
        }

    }
}

File: DLL.cpp
#include<bits/stdc++.h>

using namespace std;

struct Node{
    int data;
    Node* next = nullptr;
    Node* prev = nullptr;
};

Node* head = nullptr;

void insertAtHead(int val){
    Node* newNode = new Node();
    newNode->data = val;
    newNode->next = head;
    cout << newNode->data << "\n";
    if (head->next != nullptr){
        head->next->prev = newNode;
    }
    head = newNode;
}

void insertAtTail(int val){
    Node* cur = head;
    while (cur->next != nullptr){
        cout << cur-> data << endl;
        cur = cur->next;
    
    }
    Node* newNode = new Node();
    newNode->data = val;
    newNode->prev = cur;
    cur->next = newNode;
    
}

void insertAfter(int refval, int val){
    Node* cur = head;
    bool found = false;
    while (cur->next != nullptr && !found){
        cur = cur->next;
        if (cur->data == refval){
            found = true;
        }
    }
    Node* newNode = new Node();
    newNode->data = val;
    newNode->prev = cur;
    newNode->next = cur->next;
    newNode->next->prev = newNode;
    cur->next = newNode;
}

void displayForward(){
    Node* cur = head;
    while (cur->next != nullptr){
        cout << cur->data << " ";
        cur = cur->next;
    }
    cout << "\n";
}

void displayBackward(){
    Node* cur = head;
    while (cur->next != nullptr){
        cur = cur->next;
    }
    while (cur != head){
        cout << cur->data << " ";
        cur = cur->prev;
    }
    cout << cur->data << "\n";
}

bool search(int key){
    Node* cur = head;

    bool found = false;
    while (cur->next != nullptr && !found){
        cur = cur->next;
        if (cur->data == key){
            found = true;
        }
    }

    return found;
}


void deleteVal(int val){
    Node* cur = head;

    bool found = false;
    while (cur->next != nullptr && !found){
        cur = cur->next;
        if (cur->next->data == val){
            found = true;
        }
    }

    if (found) {
        cur->next = cur->next->next;
        cur->next->prev = cur;
        free(cur);
    }
    else cout<< "Not found" <<"\n";
}




int main(){
    insertAtHead(10);
    insertAtTail(20);
    insertAtTail(30);

    displayBackward();
}